File: src/Tests/AliasVault.UnitTests/Generators/IdentityGeneratorTest.cs
//-----------------------------------------------------------------------
// <copyright file="IdentityGeneratorTest.cs" company="lanedirt">
// Copyright (c) lanedirt. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
// </copyright>
//-----------------------------------------------------------------------

namespace AliasVault.Tests.Generators;

using System.Net.Mail;
using AliasGenerators.Identity.Implementations;
using AliasGenerators.Identity.Models;

/// <summary>
/// Tests for the CsvImportExport class.
/// </summary>
public class IdentityGeneratorTest
{
    /// <summary>
    /// Tests the generation of random dutch identities.
    /// </summary>
    /// <returns>Task.</returns>
    [Test]
    public async Task TestNlGenerator()
    {
        // Generate random NL identity.
        var generator = new IdentityGeneratorNl();

        // Generate 10 identities, check if they are all unique.
        var identities = new List<Identity>();
        for (var i = 0; i < 10; i++)
        {
            var identity = await generator.GenerateRandomIdentityAsync();
            Console.WriteLine($"Generated identity: {identity.FirstName} {identity.LastName} ({identity.Gender}) - {identity.BirthDate.ToShortDateString()} - {identity.NickName} - {identity.EmailPrefix}@domain.tld");

            identities.Add(identity);
        }

        Assert.That(identities, Has.Count.EqualTo(10), "Should generate 10 identities");
        Assert.That(
            identities,
            Is.All.Matches<Identity>(
                    i =>
                    !string.IsNullOrWhiteSpace(i.FirstName) && !string.IsNullOrWhiteSpace(i.LastName)),
            "All identities should have non-empty first and last names");
        Assert.That(identities, Is.Unique, "All identities should be unique");

        foreach (var identity in identities)
        {
            var email = identity.EmailPrefix + "@domain.tld";
            Assert.DoesNotThrow(
                () =>
                new MailAddress(email),
                $"Email address '{email}' is not a valid email. Check email prefix generation logic.");
        }
    }

    /// <summary>
    /// Tests the generation of random english identities.
    /// </summary>
    /// <returns>Task.</returns>
    [Test]
    public async Task TestEnGenerator()
    {
        // Generate random NL identity.
        var generator = new IdentityGeneratorEn();

        // Generate 10 identities, check if they are all unique.
        var identities = new List<Identity>();
        for (var i = 0; i < 10; i++)
        {
            var identity = await generator.GenerateRandomIdentityAsync();
            Console.WriteLine($"Generated identity: {identity.FirstName} {identity.LastName} ({identity.Gender}) - {identity.BirthDate.ToShortDateString()} - {identity.NickName} - {identity.EmailPrefix}@domain.tld");

            identities.Add(identity);
        }

        Assert.That(identities, Has.Count.EqualTo(10), "Should generate 10 identities");
        Assert.That(
            identities,
            Is.All.Matches<Identity>(
                i =>
                    !string.IsNullOrWhiteSpace(i.FirstName) && !string.IsNullOrWhiteSpace(i.LastName)),
            "All identities should have non-empty first and last names");
        Assert.That(identities, Is.Unique, "All identities should be unique");

        foreach (var identity in identities)
        {
            var email = identity.EmailPrefix + "@domain.tld";
            Assert.DoesNotThrow(
                () =>
                    new MailAddress(email),
                $"Email address '{email}' is not a valid email. Check email prefix generation logic.");
        }
    }
}

File: browser-extension/src/utils/generators/Identity/__tests__/IdentityGenerator.test.ts
import { IdentityGeneratorEn } from '../implementations/IdentityGeneratorEn';
import { IdentityGeneratorNl } from '../implementations/IdentityGeneratorNl';
import { describe, it, expect } from 'vitest';
import { IIdentityGenerator } from '../interfaces/IIdentityGenerator';
import { Gender } from '../types/Gender';
/**
 * Test the identity generator.
 */
const testIdentityGenerator = (
  language: string,
  generator: IIdentityGenerator
) : void => {
  describe(`IdentityGenerator${language}`, () => {
    describe('generateRandomIdentity', () => {
      it('should generate a random gender identity when no gender is specified', async () => {
        const identity = await generator.generateRandomIdentity();

        expect(identity).toBeDefined();
        expect(identity.firstName).toBeTruthy();
        expect(identity.lastName).toBeTruthy();
        expect([Gender.Male, Gender.Female]).toContain(identity.gender);
      });

      it('should generate unique identities on subsequent calls', async () => {
        const identity1 = await generator.generateRandomIdentity();
        const identity2 = await generator.generateRandomIdentity();

        expect(identity1).not.toEqual(identity2);
      });

      it('should generate an identity with all non-empty fields', async () => {
        const identity = await generator.generateRandomIdentity();

        Object.entries(identity).forEach(([, value]) => {
          expect(value).toBeTruthy();
          expect(value).not.toBe('');
          expect(value).not.toBeNull();
          expect(value).not.toBeUndefined();
        });

        // Add length checks for first and last names
        expect(identity.firstName.length).toBeGreaterThan(1);
        expect(identity.lastName.length).toBeGreaterThan(1);
      });
    });
  });
};

// Run tests for each language implementation
describe('Identity Generators', () => {
  testIdentityGenerator('En', new IdentityGeneratorEn());
  testIdentityGenerator('Nl', new IdentityGeneratorNl());
});